opentreemap/treemap/audit.py

```diff
@@ -541,7 +541,10 @@ def __init__(self, *args, **kwargs):
         # with any fields that are added during instance initialization.
         self._do_not_track = self.do_not_track
         super(UserTrackable, self).__init__(*args, **kwargs)
-        self.populate_previous_state()
+
+        # It is the leaf class' responsibility to call
```

Comment:
`populate_previous_state` calls `as_dict`, which was failing if it was called before `UDFModel.__init__` was called.

I eventually figured out that the reason for `UDFDescriptor` was to get some initialization to happen prior to `__init__`, specifically for the functioning of `as_dict`.

This strikes me as a tight coupling kind of Bad Smell. `as_dict` needs to have all initialization done, but was being called before all initialization could be done. `UDFDescriptor` was a huge headache to debug. I can't remember the exact failure modes, but setting breakpoints in it, or in the `ModelBase` method that set it up had problems. I ended up using print statements with the `traceback` module to debug it. After a while, I had enough of that, and decided to try a different approach that could eliminate `ModelBase`, `UDFDescriptor`, and the bad smell of tight coupling.

You'll know that someone failed to add a call to `populate_previous_state` because no audit records will be produced.